### PR TITLE
Introduce asio lib

### PR DESCRIPTION
**Pull request by vincent-grosbois**

### CODE REVIEW

**Comment by vincent-grosbois** [comment]
I have no idea what I need to change to have the dependencies in Linux / Mac, can someone please help?

**Comment by vincent-grosbois** [comment]
Welp, turns out that asio doesn't work on Vita/Switch because it's non-posix :(

**Comment by vincent-grosbois** [comment]
@ihhub @oleg-derevenetz could I get your feedback on this?

**Comment by oleg-derevenetz** [comment]
Hi all,

> To proceed with the task I can suggest of creating a new branch in this repository where we can approve this pull request and continue with the rest of work. Once we have needed amount of source code we can create a pull request including all changes done for networking.

I agree. To add a third-party library to a project, we need at least some minimal functionality that uses this library.

**Comment by vincent-grosbois** [comment]
ok so it's a new branch on this repo in that case ? I'm ok with it, but I don't know who has to create it

**Comment by ihhub** [comment]
Hi @vincent-grosbois , I created `network` branch here. Please assign your pull request to this branch as destination.

**Comment by ihhub** [comment]
I've changed destination branch for this pull request.

**Comment by vincent-grosbois** [comment]
thanks ! is there any objection to merge it then ? :D

File: src/fheroes2/network/test.cpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef NO_NETWORK
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wshadow" // required because asio will trigger this warning on macOS builds
+#include "asio.hpp"
+#pragma GCC diagnostic pop
+
+asio::io_context test;
+
+#endif // NO_NETWORK
```
Comment (ihhub) [inline src/fheroes2/network/test.cpp:10]: Please add an empty line here.

File: src/fheroes2/network/test.cpp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#ifndef NO_NETWORK
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wshadow" // required because asio will trigger this warning on macOS builds
+#include "asio.hpp"
+#pragma GCC diagnostic pop
+
+asio::io_context test;
```
Comment (ihhub) [inline src/fheroes2/network/test.cpp:8]: Please put this variable into anonymous namespace to reduce its scope.

File: cmake/Findasio.cmake [hunk 3]
```diff
@@ -0,0 +1 @@
+find_path(asio_INCLUDE_DIR NAMES asio.hpp)
```
Comment (ihhub) [inline cmake/Findasio.cmake:1]: Please add an empty line at the end of this file.

File: src/fheroes2/network/test.cpp [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#ifndef NO_NETWORK
```
Comment (ihhub) [inline src/fheroes2/network/test.cpp:1]: Please use `#if !defined(NO_NETWORK)` as it's easier for most of developers to miss **n**.

**Review by ihhub — COMMENTED** [review]
Hi @vincent-grosbois , overall the changes look clear. I left just few small corrections. Could you please take a look?

**Comment by ihhub** [comment]
@vincent-grosbois , well done! One step is completed!

**Pull request merged**